app/src/components/GameRoutes.tsx

```diff
@@ -15,23 +27,82 @@ function CurrentPlayerProvider(props: {
   joinCode: string
   children: React.ReactNode
 }) {
-  const { data, loading } = useCurrentPlayerQuery({
+  const currentAuth = React.useContext(CurrentAuthContext)
+
+  const [currentPlayer, setCurrentPlayer] = React.useState<
+    CurrentPlayerQuery["players"][0] | null
+  >(null)
+  const [redirectRoute, setRedirectRoute] = React.useState("")
+
+  // https://github.com/apollographql/react-apollo/issues/3505#issuecomment-568112061
+  const [skipCallback, setSkipCallback] = React.useState(false)
+
+  const [joinGame] = useJoinGameMutation()
+  const [signUp] = useLazyQuery(GameByJoinCodeDocument, {
+    variables: { joinCode: props.joinCode.toLocaleUpperCase() },
+    onCompleted: async data => {
+      if (skipCallback) {
+        return
+      }
+      if (data && data.games[0]) {
+        const registration = await joinGame({
+          variables: {
+            gameId: data.games[0].id,
+            clientUuid: clientUuid()
+          }
+        })
+        if (registration.data?.joinGame) {
+          await currentAuth.setJwtToken(registration.data.joinGame.jwt_token)
+        }
+        setSkipCallback(true)
+      } else {
+        setRedirectRoute(routes.game.root)
+      }
+    },
+    onError: _ => {
+      setRedirectRoute(routes.game.root)
+    }
+  })
+
+  const [
+    signIn,
+    { loading: loadingSignIn, called: calledSignIn }
+  ] = useLazyQuery(CurrentPlayerDocument, {
     variables: {
-      joinCode: props.joinCode,
+      joinCode: props.joinCode.toLocaleUpperCase(),
       clientUuid: clientUuid()
+    },
+    onCompleted: async data => {
+      if (data?.players[0]) {
+        setCurrentPlayer(data?.players[0])
+      }
```

Comment:
we probably want an onError callback, and to redirect to root in an else branch of the onCompleted (like above)

Reply:
e.g. this still might fail if you are trying to join a new game from the same browser (an still have your JWT for a different game)